dandiapi/api/tasks/zarr.py

```diff
@@ -0,0 +1,149 @@
+from datetime import datetime
+from typing import Optional
+
+import boto3
+from celery import shared_task
+
+from dandiapi.api.models.zarr import ZarrArchive
+from dandiapi.api.storage import get_storage
+from dandiapi.api.zarr_checksums import (
+    ZarrChecksum,
+    ZarrChecksumFileUpdater,
+    ZarrChecksumListing,
+    ZarrChecksumModification,
+    ZarrChecksumUpdater,
+)
+
+try:
+    from minio_storage.storage import MinioStorage
+except ImportError:
+    # This should only be used for type interrogation, never instantiation
+    MinioStorage = type('FakeMinioStorage', (), {})
+
+
+class SessionZarrChecksumFileUpdater(ZarrChecksumFileUpdater):
+    """Override ZarrChecksumFileUpdater to ignore old files."""
+
+    def __init__(self, *args, session_start: datetime, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        # Set the start of when new files may have been written
+        self._session_start = session_start
+
+    def read_checksum_file(self) -> Optional[ZarrChecksumListing]:
+        """Load a checksum listing from the checksum file."""
+        storage = self.zarr_archive.storage
+        checksum_path = self.checksum_file_path
+
+        # Check if this file was modified during this session
+        read_existing = False
+        if storage.exists(checksum_path):
+            read_existing = storage.modified_time(self.checksum_file_path) >= self._session_start
```

Comment:
This is to both protect the script from overwriting directories that it's already partially processed (since the max size of a object listing call is 1000 files), while ignoring any checksum entries that existed prior to this run.

If the checksum files were treated as write only, and if a directory ever contains over 1000 files, the following would occur:
* The first batch of files would be processed, and all checksums to the top of the hierarchy would be written.
* The second batch of files would be processed, but it would ignore the previously written checksum file, and so the newly written checksum file would only contain this batch of files, making the checksum file incorrect.

Reply:
Good point :+1: